corehq/apps/hqadmin/service_checks.py

```diff
@@ -222,6 +223,42 @@ def check_formplayer():
         return ServiceStatus(res.ok, msg)
 
 
+def check_pillowtop():
+    def changes_behind(pillow_meta):
+        lag = 0
+        for k, offset in pillow_meta['offsets'].items():
+            # seq_format is 'json' for KafkaChangeFeed and 'text' for
+            # ChangeFeed. The former seems to have offsets that are greater
+            # than the corresponding seq, while the latter has a seq value
+            # that's greater than the offset
+            if pillow_meta['seq_format'] == 'json':
+                seq = pillow_meta['seq'].get(k, 0)
+                lag += offset - seq
+            else:
+                seq = pillow_meta['seq']
+                lag += seq - offset
+        return lag
+
+    bad_pillows = []
+    for pillow_meta in get_all_pillows_json():
+        # numbers chosen arbitrarily
+        if (changes_behind(pillow_meta) > 100
+                # More than two days without a checkpoint?  Probably intentionally inactive
+                and 60 * 60 * 48 > pillow_meta['seconds_since_last'] > 60 * 10):
```

Comment:
It looks like pillows only checkpoint upon processing a change that occurs at least 300 seconds after the last checkpoint.  This check isn't ideal, since it's conceivable that a pillow would receive a bunch of changes within the first 300 seconds, then nothing, so it'd show a lag, but there wouldn't actually be changes to process.

Reply:
I've been wondering how to handle this as well. Ideally we want to know exactly which pillows are expected to be running on an environment. We have the info in the [app-processes.yml](https://github.com/dimagi/commcare-cloud/blob/master/environments/production/app-processes.yml#L78) so maybe we can just write it into localsettings?

Reply:
I've thought a bit about this. There's a timeout parameter on the kafka consumer class we use in change feeds, that should throw an exception after a timeout without changes has occurred. I think it could look like:

```python
try:
    for change in kafka_procuder.changes(timeout=300):
       process_change()
except KafkaTimeout:
    force_checkpoint()
finally:
    keep_processing()
```

I recall looking at this before and there was some weird issue I ran into. I think we could do something similar for couch by removing the heartbeat from couch changes

Reply:
That's a good point, @snopoke .  @emord just pointed out that the [LOCAL_PILLOWS](https://github.com/dimagi/commcare-cloud/blob/master/environments/icds-new/public.yml#L177) setting currently defines additional pillows for an environment, beyond those specified in settings.py.  We could shuffle things around such that:

* settings.py contains a list of the only basic pillows that are expected to run on every environment (right now it has some custom stuff too)
* localsettings.py contains a list of pillows that are run only on a subset of environments

Then we would be able to conclusively tell from `settings` which pillows should or shouldn't be running.

The second part is knowing whether or not the process is still alive.  It sounds like we shouldn't use `all_pillows_supervisor_status` (it would be difficult to do anyways).  I like Emord's suggestion of making the pillow process perform a checkpoint every 5 minutes whether or not there's a change that comes through.

Not sure whether I'll get to either or both of those changes any time soon, but I wanted to note them down here as next steps.

Reply:
I had forgotten that we actually have `ACTIVE_PILLOW_NAMES` which lists all the pillows that are supposed to be running. We already use this in a few places to filter the full list of pillows

Reply:
Oh nice, looks like that's been incorporated into `get_all_pillows_json`, which this uses to fetch the list.  I think this still needs a way to tell conclusively whether the pillow is receiving changes and not processing them.

Reply:
I think the only way to do that would be to use a 'heartbeat' approach like we do for celery and then define acceptable delays before a blockage is reported.

Reply:
Unless we keep track of checkpoint changes over time so that we can detect if they aren't getting updated.